src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -174,16 +172,106 @@ private sealed class RunningInstance
+                            Interop.EventStream.FSEventStreamScheduleWithRunLoop(eventStream, s_watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+                            return;
+                        }
 
-                        // Stop the FS event message pump
-                        Interop.RunLoop.CFRunLoopStop(_watcherRunLoop);
+                        s_scheduledStreamsCount = 1;
+                        var runLoopStarted = new ManualResetEventSlim();
+                        new Thread(WatchForFileSystemEventsThreadStart) { IsBackground = true }.Start(new object[] { runLoopStarted, eventStream });
+                        runLoopStarted.Wait();
+                    }
+                }
+
+                public static void UnscheduleFromRunLoop(SafeEventStreamHandle eventStream)
+                {
+                    if (s_watcherRunLoop != IntPtr.Zero)
+                    { 
+                        lock (s_lockObject)
+                        {
+                            if (s_watcherRunLoop != IntPtr.Zero)
+                            { 
+                                // Always unschedule the RunLoop before cleaning up
+                                Interop.EventStream.FSEventStreamUnscheduleFromRunLoop(eventStream, s_watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+                                s_scheduledStreamsCount--;
+
+                                if (s_scheduledStreamsCount == 0)
+                                {
+                                    // Stop the FS event message pump
+                                    Interop.RunLoop.CFRunLoopStop(s_watcherRunLoop);                                    
+                                    s_watcherRunLoop = IntPtr.Zero;
+                                }
+                            }
+                        }
+                    }
+                }
+
+                private static void WatchForFileSystemEventsThreadStart(object args)
+                {
+                    var inputArgs = (object[])args;
+                    var runLoopStarted = (ManualResetEventSlim)inputArgs[0];
+                    var _eventStream = (SafeEventStreamHandle)inputArgs[1];
+                    // Get this thread's RunLoop
+                    var runLoop = Interop.RunLoop.CFRunLoopGetCurrent();
+                    s_watcherRunLoop = runLoop;
+                    Debug.Assert(s_watcherRunLoop != IntPtr.Zero);
+
+                    // Retain the RunLoop so that it doesn't get moved or cleaned up before we're done with it.
+                    IntPtr retainResult = Interop.CoreFoundation.CFRetain(runLoop);
+                    Debug.Assert(retainResult == runLoop, "CFRetain is supposed to return the input value");
+
+
+                    // Schedule the EventStream to run on the thread's RunLoop
+                    Interop.EventStream.FSEventStreamScheduleWithRunLoop(_eventStream, runLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
```

Comment:
I'm still not clear on why the first event stream is special.  Is it that CFRunLoopRun requires at least one event stream scheduled on it already in order to work?  Otherwise, it'd be better if the ScheduleEventStream method were the only call site to FSEventStreamScheduleWithRunLoop.

Reply:
RunLoop won't run without any scheduled event stream.